### PR TITLE
NPC Sleeping Zone

### LINKED ISSUE #71865: NPCs have trouble finding good sleeping spots and staying asleep
### Is your feature request related to a problem? Please describe.

Currently managing NPC sleep is more frustrating than it needs to be - different followers go to sleep at different times, and when they do, they seek the nearest seat or bed and go right to sleep. They are constantly awakened by noises, such as walking, stereo, or even other followers saying "Bye" at 15 sound whenever you talk to them. Other times, NPCs even take up your spot in the bed for themselves.

If you want to assign them beds to sleep in, you have to herd them one by one to a desired spot, command them to sleep, and wait for them to lie down. This is so many button presses just to be unbothered during your overnight crafting session.

### Solution you would like.

~~A new zone in the zone manager to define where NPCs will go to sleep. Followers will seek beds and seats within the zone and use them.~~

~~It would also be nice if followers automatically crouched while moving about within the sleep zone as to try not to wake eachother up, but this wouldn't be 100% necessary.~~

NPCs should search a wider area when attempting to go to sleep and choose comfort over proximity if they can. The player can then manage where NPCs go to sleep by simply furnishing the bedroom(s) and making them desirable places to sleep. 

Doing it this way instead of zones means that the player shouldn't need to define where their followers should rest every time they have to camp somewhere outside their usual base. 

NPCs should also have a memory of recent sleeping spots and prefer to stick to that, or at least avoid taking the player character's bed unless there's no other option.

### Describe alternatives you have considered.

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Norwaivhi**
#### Summary
None

#### Purpose of change
To add a suggested feature, that gives an option of manually setting preferred sleeping spaces for friendly NPCs to the Player;
To close #71865 

#### Describe the solution
New zone type - Sleep Zone;
You can set it using the zone manager (Y), and after doing so, friendly NPCs, while within the camp limits, will first look for rest spots in those designated zones, choosing the best one.
If no such spots are found, the code falls back to the old algorithm of finding the best resting spot nearby.

Added new zone type to the Zones.json;
Added a new line to the TALK_FACTION_CAMP.json to mention the new zone type;

#### Describe alternatives you've considered
At first, I was thinking of "sleeping spots" that would contain only one tile, this way checks would be faster, but zone is not just one tile and it would be quite messy to have an individual zone for each bed;

#### Testing
This exact feature is working as expected:
(The biggest problem is other NPCs screaming and being loud, which makes testing with multiple NPCs and areas a bit cumbersome, but it still does show expected behavior)
Zones can be set with zone manager, they save and load without any problems;
NPCs consider zoned areas to sleep before picking the best they can find;

#### Additional context
Old code is left in and it triggers by default if the feature is not been used, so all should be perfectly fine if Players won't notice or use it;
The zone name and description might need to be changed, I'm not sure what is the best language there;

### CODE REVIEW

**Comment by RenechCDDA** [comment]
I think that the linked issue has missed the point and *more micromanagement* is not a solution. If there are individual problems with NPC interactions (all phrases being shouted at max volume which wakes everyone up etc) those should be individually addressed.

NPCs are already able to seek out beds on their own so I am not sure what a specifically defined sleep zone adds, even. It gives a further level of power to the player over the already-too-subservient NPCs and encourages the player with the expectation that they *should* micromanage their NPCs. They should not. NPCs should be capable of finding beds on their own (and can, as noted!).

File: src/npcmove.cpp [hunk 1]
```diff
@@ -1638,23 +1638,61 @@ void npc::execute_action( npc_action action )
...
+            int is_in_camp = within_boundaries_of_camp() ? 1 : 0;
+            switch( is_in_camp ) {
+                case 1:
+                    if( is_friendly( player_character ) ) {
+                        zone_manager &zone_manager = zone_manager::get_manager();
+                        faction *player_faction = player_character.get_faction();
+                        zone_type_id zone_type( "SLEEP_ZONE" );
+                        std::vector<const zone_data *> zone_info = zone_manager.get_near_zones( zone_type,
+                                get_location(),
+                                120,
```
Comment (RenechCDDA) [inline src/npcmove.cpp:1651]:
```suggestion
                                ACTIVITY_SEARCH_DISTANCE,
```

**Comment by Maleclypse** [comment]
One of your statements left me with a question.  Is there any weird interaction if there are multiple sleeping zones in an area.  Like If my camp has four bedrooms and each bedroom has a sleep zone covering it I would have four sleep zones.  Assuming I have 8 NPCs will they all try to sleep in one zone and be mad it's full up?

**Comment by Norwaivhi** [comment]
The code looks for all sleeping zones in a range and checks every tile of those found for comfort;
It does not implicitly force NPCs to sleep there, but points to the place where to look for beds first;

It picks a spot if:
- It is better than the current one
- Is passable(Not a wall for example)
- Is empty(Not occupied by player\monster\npc)

Let's imagine a default evac shelter camp, with a few bedrooms built(example from my savefile):

I have 4 NPCs and 2 bedrooms built, 8 beds total;

1. The first NPC goes to spot 1-> 
2. Because it is no longer empty the Second NPC goes to spot 2->
3. Because it is no longer empty the Third NPC goes to spot 3->
4. The process continues until all NPCs are in beds or all beds are taken;

If we end up with all decent spots in the zones taken the switch in the code falls through and reverts to the old algorithm that looks for the most comfortable space nearby. In that case (Imagine we have like 20 NPCs), it would start looking for the best spots in the "living room" and that would be those sofas I'm standing on, armchairs, chairs, and so on.

It should not have many issues with scaling amount of zones and NPCs, other than taking longer to check all the tiles if the zones are huge.

Does that answer your question?

**Comment by kevingranade** [comment]
No thanks,  I agree with Renech that more micromanagement over NPCs is not a viable answer here.

**Comment by esotericist** [comment]
this isn't how we want faction camps to work. the extent of the player's involvement in beds should, at most, to be to go "this one is mine", and cause NPCs to prefer beds other than that.

**Comment by kevingranade** [comment]
On top of that,  I'd say that claim should be implicit,  if you sleep in a bed it should claim it,  and NPCS will then avoid it unless they don't have an alternative.

**Pull request closed without merging**